### PR TITLE
+ ECE + discontinuous regression

### PR DESCRIPTION
**Pull request by tennisonliu**
* + expected calibration error calculations
* - remove saved models
* + regression with artificial gap
* clean up tracking
* update README with extensions

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,6 +7,11 @@ Paper: Blundell, C., Cornebise, J., Kavukcuoglu, K. and Wierstra, D., 2015, June
         In International Conference on Machine Learning (pp. 1613-1622). PMLR.
        (http://proceedings.mlr.press/v37/blundell15.html)
 
+## Extension
+Additional approximate inference methods are implemented including:
+* *Bayes by Backprop - Local Reparameterisation Trick* (https://arxiv.org/pdf/1506.02557.pdf): Samples pre-activations instead of weights for lower variance, faster computation and convergence.
+* *Monte Carlo Dropout* (https://arxiv.org/pdf/1506.02142.pdf): Dropout during test-time to generate uncertainty measures, p=0.5.
+* *Functional Variational Inference* (https://arxiv.org/pdf/1903.05779.pdf): Optimises ELBO defined on stochastic processes, *i.e.* distribution over functions.
```
Comment (chelsealuisa) [inline README.md:14]: Nice!

File: README.md [hunk 2]
```diff
@@ -23,5 +28,10 @@ At run-time, `main` reads from a model configuration set in `config.py`. The con
 python3 main.py --model [regression|classification|rl]
 ```
 
+## Analysis:
+The scripts `weight_pruning.py` and `compute_ece.py` perform post-hoc analysis using saved models.
+* `weight_pruning` 1) plots the distribution of weights, 2) computes SNR of BNNs, 3) evaluates performance on pruned weights.
+* `compute_ece` 1) computes the ECE of trained model, 2) plots reliability diagram.
```
Comment (chelsealuisa) [inline README.md:34]: Stupid question - what does ECE stand for?

File: classification/class_task.py [hunk 3]
```diff
@@ -165,7 +165,7 @@ def evaluate(self, test_loader):
             for data in tqdm(test_loader):
                 X, y = data
                 X, y = X.to(DEVICE), y.to(DEVICE)
-                preds = self.predict(X)
+                preds, _ = self.sample_predict(X)
```
Comment (chelsealuisa) [inline classification/class_task.py:168]: Possibly way too nit-picky, but should we instead change all of the sample_predict methods to just predict()? sample_predict doesn't really make sense for MLP

File: config.py [hunk 4]
```diff
@@ -13,12 +13,12 @@ class RegConfig:
...
     epochs = 1000
     train_samples = 5
-    test_samples = 100
+    test_samples = 10
     num_test_points = 400
     mode = 'regression'
     mixture_prior = False
-    hidden_units = 400
-    noise_tolerance = .1
+    hidden_units = 600              # CHANGED TO 100
```
Comment (chelsealuisa) [inline config.py:20]: Could we leave the config settings as they were to reflect the values that were used to create the final saved models/results that will be included in the report?

File: compute_ece.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    models = {
+        'BNN': BNN_Classification('bnn_classification', {**params, 'local_reparam': False, 'dropout': False}),
+        'BNN-LR': BNN_Classification('bnn_classification_lr', {**params, 'local_reparam': True, 'dropout': False}),
+        'MLP': MLP_Classification('mlp_classification', {**params, 'dropout': False}),
+        # 'dropout_class': MLP_Classification('dropout_classification', {**params, 'dropout': True}),
+        'MC-Dropout': MCDropout_Classification('mcdropout_classification', {**params, 'dropout': True}),
+        }
+
+    for _, model in models.items():
+        model.net.load_state_dict(torch.load(model.save_model_path))
```
Comment (chelsealuisa) [inline compute_ece.py:113]: Could this be replaced with the functions in load_model_utils instead?

File: load_model_utils.py [hunk 6]
```diff
@@ -1,78 +1,9 @@
...
-            log_variational_posteriors[i] = self.log_variational_posterior()
-            negative_log_likelihood += self.get_nll(output, target, sigma)
-
-        log_prior = beta*log_priors.mean()
-        log_variational_posterior = beta*log_variational_posteriors.mean()
-        negative_log_likelihood = negative_log_likelihood / samples
-        loss = log_variational_posterior - log_prior + negative_log_likelihood
-        return loss, log_priors.mean(), log_variational_posteriors.mean(), negative_log_likelihood
-
-def load_bnn_class_model(saved_model):
```
Comment (chelsealuisa) [inline load_model_utils.py:75]: thx for removing. This was a temporary addition to be able to work with the older saved models that had different model params, I should've removed it already.

File: main.py [hunk 7]
```diff
@@ -13,7 +13,7 @@
 def reg_trainer():
     ''' Regression Task Trainer '''
     config = RegConfig
-    X, Y = create_data_reg(train_size=config.train_size)
+    X, Y = create_data_reg(train_size=config.train_size, gap=True)
```
Comment (chelsealuisa) [inline main.py:16]: Shouldn't this be set to False as a default?

File: README.md (hunk 2 above)
Comment (tennisonliu) [reply README.md:34]: done

File: classification/class_task.py (hunk 3 above)
Comment (tennisonliu) [reply classification/class_task.py:168]: done.

File: config.py (hunk 4 above)
Comment (tennisonliu) [reply config.py:20]: done.

File: compute_ece.py (hunk 5 above)
Comment (tennisonliu) [reply compute_ece.py:113]: That's what I thought to do, but we need functions e.g. predict() and enable_dropout() that are only available in the class_task objects and not the models.

File: main.py (hunk 7 above)
Comment (tennisonliu) [reply main.py:16]: moved this into configuration file.

**Comment by tennisonliu** [comment]
@chelsealuisa made the following changes:
1) uploaded (final) plots in graphs/final_graphs
2) removed saved models from repo
3) updated readme with poster link
4) moved util functions into /utils folder (along with necessary refactoring to change relative paths)

**Comment by chelsealuisa** [comment]
Looks great!

**Review by chelsealuisa — APPROVED** [review]
